hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java

```diff
@@ -219,123 +219,140 @@ protected void onPutIfAbsentAsyncInternal(K key, V value, Data keyData, Data val
+    private void populateResultFromRemote(int currentSize, List<Object> resultingKeyValuePairs, Map<Object, Long> reservations) {
+        for (int i = currentSize; i < resultingKeyValuePairs.size(); i += 2) {
+            Data keyData = (Data) resultingKeyValuePairs.get(i);
+            Data valueData = (Data) resultingKeyValuePairs.get(i + 1);
+
+            Object ncKey = serializeKeys ? keyData : toObject(keyData);
+            if (!serializeKeys) {
+                resultingKeyValuePairs.set(i, ncKey);
+            }
+
+            Long reservationId = reservations.get(ncKey);
+            if (reservationId != null) {
+                Object cachedValue = tryPublishReserved(ncKey, valueData, reservationId);
+                resultingKeyValuePairs.set(i + 1, cachedValue);
+                reservations.remove(ncKey);
             }
         }
     }
 
     @Override
-    protected void putAllInternal(Map<? extends K, ? extends V> map, ExpiryPolicy expiryPolicy,
+    protected void putAllInternal(Map<? extends K, ? extends V> map, ExpiryPolicy expiryPolicy, Map<Object, Data> keyMap,
                                   List<Map.Entry<Data, Data>>[] entriesPerPartition, long startNanos) {
         try {
-            super.putAllInternal(map, expiryPolicy, entriesPerPartition, startNanos);
-            cacheOrInvalidate(map, entriesPerPartition, true);
+            if (!serializeKeys) {
+                keyMap = createHashMap(map.size());
+            }
+            super.putAllInternal(map, expiryPolicy, keyMap, entriesPerPartition, startNanos);
+            cacheOrInvalidate(map, keyMap, entriesPerPartition, true);
         } catch (Throwable t) {
-            cacheOrInvalidate(map, entriesPerPartition, false);
+            cacheOrInvalidate(map, keyMap, entriesPerPartition, false);
             throw rethrow(t);
         }
     }
 
-    private void cacheOrInvalidate(Map<? extends K, ? extends V> map, List<Map.Entry<Data, Data>>[] entriesPerPartition,
-                                   boolean isCacheOrInvalidate) {
+    private void cacheOrInvalidate(Map<? extends K, ? extends V> map, Map<Object, Data> keyMap,
+                                   List<Map.Entry<Data, Data>>[] entriesPerPartition, boolean isCacheOrInvalidate) {
         if (serializeKeys) {
             for (int partitionId = 0; partitionId < entriesPerPartition.length; partitionId++) {
                 List<Map.Entry<Data, Data>> entries = entriesPerPartition[partitionId];
                 if (entries != null) {
                     for (Map.Entry<Data, Data> entry : entries) {
+                        Data key = entry.getKey();
                         if (isCacheOrInvalidate) {
-                            cacheOrInvalidate(entry.getKey(), null, entry.getValue());
+                            // FIXME: the null value produces an unwanted deserialization if Near Cache in-memory format is OBJECT
+                            cacheOrInvalidate(key, key, null, entry.getValue());
```

Comment:
This is the remaining code part, where we eventually deserialize the value twice in a distinct use case.